### PR TITLE
Migrate Domain Faceted Report to Bootstrap 3

### PR DESCRIPTION
**Pull request by biyeun**
presenting a report, in the terrible old framework, migrated to bootstrap 3 with minimal refactoring.
@orangejenny @sravfeyn @czue @benrudolph 

cc @dimagi/product (hehe)

screen shot 2015-12-29 at 7 00 11 pm
screen shot 2015-12-29 at 7 00 05 pm
screen shot 2015-12-29 at 6 59 56 pm

### CODE REVIEW

File: custom/_legacy/mvp/reports/chw.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ class CHWManagerReport(GenericTabularReport, MVPIndicatorReport, DatespanMixin):
     @memoized
     def template_report(self):
         if self.is_rendered_as_email:
-            self.report_template_path ="reports/async/tabular.html"
+            self.report_template_path ="reports/async/bootstrap2/tabular.html"
```
Comment (dimagimon) [inline custom/_legacy/mvp/reports/chw.py]: E225 missing whitespace around operator

File: corehq/apps/style/static/style/less/bootstrap3/report.less [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+.report-loading-container {
+  position: relative;
+}
+.report-loading {
+  position: absolute;
+  background-color: #ffffff;
+  padding: 15px 20px;
+  z-index: 100;
```
Comment (orangejenny) [inline corehq/apps/style/static/style/less/bootstrap3/report.less]: how about putting a variable for this in variables.less? why is this 100, what does it need to be above/below?

File: corehq/apps/appstore/static/appstore/js/facet_sidebar.js [hunk 3]
```diff
@@ -88,4 +97,4 @@ $(function () {
         e.preventDefault();
         $update_btn.click();
     });
-});
\ No newline at end of file
+});
```
Comment (orangejenny) [inline corehq/apps/appstore/static/appstore/js/facet_sidebar.js]: Odd that this file lives in appstore when it's used outside of the exchange. Maybe hqwebapp would be better?

(Not blocking, I agree it makes sense to limit this PR to B3-necessary changes, but it would be unfortunate to review this large a chunk of code and not comment on possible future improvements.)

File: corehq/apps/appstore/templates/appstore/partials/bootstrap3/sidebar.html [hunk 4]
```diff
@@ -0,0 +1,75 @@
+{% load i18n %}
+{% load hq_shared_tags %}
+<ul class="nav nav-list nav-hq-sidebar">
```
Comment (orangejenny) [inline corehq/apps/appstore/templates/appstore/partials/bootstrap3/sidebar.html]: `.nav-list` isn't in B3...?

File: corehq/apps/appstore/templates/appstore/partials/bootstrap3/sidebar.html [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                        </label>
+                    </div>
+                    <div class="col-sm-2"><input id="s-{{ f.facet }}-{{ choice.name }}" name="{{ choice.name|urlencode }}" class="facet-checkbox" type="checkbox" {% if choice.active %}checked=checked{% endif %}/></div>
+                </div>
+
+
+            </li>
+        {% endfor %}
+        {% if f.choices|length > 6 and f.expanded %}
+            <li style="list-style-type: none;"><a class="more-sortable-button" data-sortable="{{ f.facet }}" href="#">{% trans "More..." %}</a></li>
```
Comment (orangejenny) [inline corehq/apps/appstore/templates/appstore/partials/bootstrap3/sidebar.html]: Can you use `.list-unstyled`, or does it interfere with desired margin/padding?

File: corehq/apps/hqadmin/static/hqadmin/js/visualizations.js [hunk 6]
```diff
@@ -99,7 +99,8 @@ var HQVisualizations = function (options) {
             $(self.chart_tabs_id).on('click', 'a[data-toggle="hash-tab"]', function(){
                 $('.nvd3-chart').hide();
                 var $chart = $($(this).attr('href')).children('.nvd3-chart');
-                $chart.show();
+                $chart.show().removeClass('hidden');
```
Comment (orangejenny) [inline corehq/apps/hqadmin/static/hqadmin/js/visualizations.js]: would be a good future refactor to not have to both use show/hide and toggle `.hidden`

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html [hunk 7]
```diff
@@ -0,0 +1,36 @@
+{% load hq_shared_tags %}
+{% load i18n %}
+
+<div class="hq-accordion-control">
```
Comment (orangejenny) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: Confused by this - B3 doesn't have the `.accordion-foo` classes anymore, is this an hq-specific widget?

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        {% endif %}
+        <div id="extra-filter-info" class="accordion-extra"></div>
+    </div>
+    {% if show_dates %}
+        <div id="reportFilters" class="accordion-body collapse">
+            <div class="accordion-inner">
+                <form method="get" id="paramSelectorForm" class="form-horizontal">
+                    <fieldset>
+                        <legend>Report Filters</legend>
+                        <div id="datespan" class="control-group">
```
Comment (orangejenny) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: control-group => form-group

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+                            <label class="control-label">Date Range</label>
+                            <div class="controls">
+                                <label for="startdate" class="sub-label">From</label>
+                                <input type="text" id="startdate" name="startdate" class="date-picker input-small" value="{{ datespan.startdate|date:"Y-m-d" }}">
```
Comment (orangejenny) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: `.input-small` is deprecated

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+                                <label for="enddate" class="sub-label">To</label>
+                                <input type="text" id="enddate" name="enddate" class="date-picker input-small" value="{{ datespan.enddate|date:"Y-m-d" }}">
+                                <span class="helptext help-inline" style="margin-left: 1.5em;">Showing {{ datespan }}.</span>
+                            </div>
+                        </div>
+                    </fieldset>
+                    <div class="form-actions">
+                        <button type="submit" class="btn btn-large btn-primary">Update Filters</button>
+                        <a href="#reportFilters" data-toggle="collapse" class="btn btn-large">Cancel</a>
```
Comment (orangejenny) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: needs `.btn-default`, and `.btn-large` should be `.btn-lg`

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/partials/visualizations.html [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            <div style="height: 320px; min-height: 320px;" class="loading report-loading-container">
+                <div class="report-loading{% if hq_loading %} hq-loading{% endif %}">
+                    <h4>{% trans "Loading Bar Chart" %}</h4>
+                    <img src="{% new_static 'hqwebapp/img/ajax-loader.gif' %}" alt="loading indicator" />
+                </div>
+            </div>
+            <div class="alert alert-warning no-data hidden">
+                <h4>{% trans "Your query returned no data" %}</h4>
+            </div>
+            <div class="alert alert-error error hidden">
```
Comment (orangejenny) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/partials/visualizations.html]: alert-error => alert-danger (couple of places in this file)

File: corehq/apps/hqwebapp/static/hqwebapp/js/hash-tab.js [hunk 12]
```diff
@@ -11,4 +11,9 @@ $(function(){
         window.location.hash = this.hash;
         $('html,body').scrollTop(scrollmem);
     });
+
+    $('.nav-tabs a[data-toggle="tab"]').on('shown.bs.tab', function (e) {
+        console.log('shown');
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/hash-tab.js]: remove

File: corehq/apps/reports/templates/reports/async/bootstrap3/default.html [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        </div>
+        <div class="modal-body">
+            <p>There seems to be a problem loading data for the <strong>{{ report.title }}</strong> report.</p>
+            <p>The report returned the following error:</p>
+            <div class="alert alert-error report-error-status"></div>
+            <p>If this problem persists, we encourage you to <a href="#modalReportIssue" data-target="#modalReportIssue" data-toggle="modal">Report an Issue</a>.</p>
+        </div>
+        <div class="modal-footer">
+            <a href="#" class="btn btn-primary" onclick="$(this).parent().parent().trigger('reloadHQReport')" data-loading-text="Trying Again..." data-fail-text="Loading Failed. Try Again?">Try Again</a>
+            <a href="#" class="btn" data-dismiss="modal">Cancel</a>
```
Comment (orangejenny) [inline corehq/apps/reports/templates/reports/async/bootstrap3/default.html]: btn-default

File: corehq/apps/reports/templates/reports/async/bootstrap3/default.html [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+{% block js-inline %}
+{% endblock %}
+
+{% block modals %}
+    <div class="loading-backdrop hide"></div>
+
+    <div class="modal hide fade" id="loadingReportIssueModal">
+        <div class="modal-header">
+            <a class="close" data-dismiss="modal">&times;</a>
+            <h3>Problem Loading Report "{{ report.title }}"</h3>
```
Comment (orangejenny) [inline corehq/apps/reports/templates/reports/async/bootstrap3/default.html]: translations (throughout this file)

File: corehq/apps/appstore/templates/appstore/partials/bootstrap3/sidebar.html [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+        {% endif %}
+        <li class="nav-divider"></li>
+    {% endif %}
+
+    <li class="nav-input">
+        <button class="btn btn-primary" id="update-facets" data-prefix="{{ facet_prefix }}" data-params="{{ persistent_params }}">{% trans "Apply" %}</button>
+    </li>
+    <li class="nav-input">
+        <form action="{{ search_url }}" method="get" class="form-search" id="facet-search">
+            <input type="search" name="{{ search_name }}" placeholder="Search for {{ search_for }}" value="{{ search_query }}" class="search-query form-control" />
```
Comment (orangejenny) [inline corehq/apps/appstore/templates/appstore/partials/bootstrap3/sidebar.html]: translate placeholder?

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+            <div class="no-filters">&nbsp;</div>
+        {% endif %}
+        <div id="extra-filter-info" class="accordion-extra"></div>
+    </div>
+    {% if show_dates %}
+        <div id="reportFilters" class="accordion-body collapse">
+            <div class="accordion-inner">
+                <form method="get" id="paramSelectorForm" class="form-horizontal">
+                    <fieldset>
+                        <legend>Report Filters</legend>
```
Comment (orangejenny) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: translations (throughout this file)

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_report.html [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    <div class="report-actions">
+        {% if report.is_exportable %}
+            <a href="#" class="btn btn-default{% if report.needs_filters %} hide{% endif %}" id="export-report-excel">
+                <i class="fa fa-share"></i> {% trans "Export to Excel" %}
+            </a>
+        {% endif %}
+
+        {% if report.is_emailable %}
+            <a href="#email-report-modal" class="btn{% if report.needs_filters %} hide{% endif %}" data-toggle="modal" id="email-report">
+                <i class="icon icon-envelope"></i> {% trans "Email report" %}
```
Comment (orangejenny) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_report.html]: shouldn't this be an fa icon? same with icon-print below.

File: corehq/apps/reports/templates/reports/async/bootstrap3/default.html [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+{% block js %}
+{% endblock %}
+
+{% block js-inline %}
+{% endblock %}
+
+{% block modals %}
+    <div class="loading-backdrop hide"></div>
+
+    <div class="modal hide fade" id="loadingReportIssueModal">
```
Comment (orangejenny) [inline corehq/apps/reports/templates/reports/async/bootstrap3/default.html]: this doesn't look like a B3 modal

File: locale/en/LC_MESSAGES/django.po [hunk 19]
```diff
@@ -1875,7 +1875,7 @@ msgid "Go to All Invoices page for this Project Space"
 msgstr ""
 
 #: corehq/apps/accounting/templates/accounting/partials/invoice_table.html:8
-#: corehq/apps/hqadmin/templates/hqadmin/hqadmin_base_filters.html:9
+#: corehq/apps/hqadmin/templates/hqadmin/bootstrap2/hqadmin_base_filters.html:9
```
Comment (orangejenny) [inline locale/en/LC_MESSAGES/django.po]: Curious why these changes are necessary, I thought the .po files are automatically generated...?

**Comment by orangejenny** [comment]
Looks great overall!

Build failure real?

AssertionError: new_static not found in '    <script type="text/javascript" src="{% static \'reports/javascripts/config.dataTables.bootstrap.js\' %}"></script>\n' in file /home/travis/build/dimagi/commcare-hq/corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_report.html

File: corehq/apps/appstore/templates/appstore/partials/bootstrap3/sidebar.html (hunk 5 above)
Comment (biyeun) [inline corehq/apps/appstore/templates/appstore/partials/bootstrap3/sidebar.html]: a lot of that came from the copy over I'll remove it

File: corehq/apps/hqadmin/static/hqadmin/js/visualizations.js (hunk 6 above)
Comment (biyeun) [inline corehq/apps/hqadmin/static/hqadmin/js/visualizations.js]: right, but more than just this b3 page uses this file. so I left the original functionality there since it doesn't break anything (it just doesn't apply)

Comment (orangejenny) [inline corehq/apps/hqadmin/static/hqadmin/js/visualizations.js]: ah, i didn't realize this was a shared file. cool.

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html (hunk 7 above)
Comment (biyeun) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: again, this is from the copied part of the file. It might be good to review commit-by-commit to see what I updated? This isn't shown on this page, so I didn't work on updating the styles yet. I'm saving that for a page that actually uses filters...keeping the changes here as basic as possible

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html (hunk 8 above)
Comment (biyeun) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: yeah, again...I didn't touch this because it wasn't visible.

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html (hunk 9 above)
Comment (biyeun) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: see above note

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html (hunk 10 above)
Comment (biyeun) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: see above note

File: corehq/apps/reports/templates/reports/async/bootstrap3/default.html (hunk 18 above)
Comment (biyeun) [inline corehq/apps/reports/templates/reports/async/bootstrap3/default.html]: thanks, I actually missed this one

File: corehq/apps/appstore/static/appstore/js/facet_sidebar.js (hunk 3 above)
Comment (biyeun) [inline corehq/apps/appstore/static/appstore/js/facet_sidebar.js]: i'm going to leave this one be

File: corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html (hunk 7 above)
Comment (orangejenny) [inline corehq/apps/hqadmin/templates/hqadmin/bootstrap3/hqadmin_base_filters.html]: Hrm, it's odd to me to have files that live in a bootstrap3 directory but aren't migrated. Don't think it's a huge deal, but a little confusing.

Yeah, I only read the combined diff, definitely feel free to ignore comments on stuff you didn't touch that would be distracting to deal with right now.

File: custom/_legacy/mvp/reports/chw.py (hunk 1 above)
Comment (dimagimon) [inline custom/_legacy/mvp/reports/chw.py]: E225 missing whitespace around operator

**Comment by biyeun** [comment]
@orangejenny responded! thanks for catching all of those! A few of the issues I was aware of and purposely left for a later date as they didn't affect this PR and would need to be addressed for a migration of other reports (and I couldn't see the effects reliably anyway). Commit-by-commit is a good one to see what bits of code were left over from the original templates. ups to @benrudolph for the `new_static` test :smiley:

File: locale/en/LC_MESSAGES/django.po (hunk 19 above)
Comment (biyeun) [inline locale/en/LC_MESSAGES/django.po]: yeah, I figured it was fine to update regardless

**Comment by orangejenny** [comment]
sweet. noticed one more thing - this should be alert-danger: https://github.com/dimagi/commcare-hq/commit/93323aa74407fb9443fd3270aea8b53049156768?w=1#diff-77619115b549d0fa19cab27e9d2501c3R61

otherwise, good to merge once tests pass!

File: corehq/apps/style/tests/test_compress_command.py [hunk 20]
```diff
@@ -32,6 +37,8 @@ def _assert_valid_import(self, line, filename):
         )
 
     def _is_b3(self, filename):
+        if filename in IGNORED_FILES:
+            return  False
```
Comment (dimagimon) [inline corehq/apps/style/tests/test_compress_command.py]: E271 multiple spaces after keyword

File: custom/_legacy/mvp/reports/chw.py (hunk 1 above)
Comment (dimagimon) [inline custom/_legacy/mvp/reports/chw.py]: E225 missing whitespace around operator

**Comment by biyeun** [comment]
updates @orangejenny

**Comment by czue** [comment]
wow, looks like it was a battle! great to see this!
